torch/ao/quantization/dynamic_tracing/quantization_state.py

```diff
@@ -0,0 +1,725 @@
+        """
+
+        if isinstance(outputs, torch.Tensor):
+            qtensor_info = self.output_qtensor_infos[0]
+            if self.output_dtypes is not None:
+                assert qtensor_info is not None
+                # check the output dtype, and do the conversion if needed
+                output_dtype = self.output_dtypes[0]
+                if qtensor_info.inf_dtype != output_dtype:
+                    assert output_dtype is torch.float, \
+                        'non-float output dtypes not handled yet'
+                    outputs = outputs.dequantize()
+            else:
+                # if no output dtype was specified, do nothing
+                pass
+
+        return outputs
+
+    def get_output_qtensor_infos(self) -> List[Optional[QTensorInfo]]:
+        """
+        Used by the conversion to torch.jit.script.
+        """
+        return self.output_qtensor_infos
+
+    def get_output_dtypes(self) -> Any:
+        """
+        Used by the conversion to torch.jit.script.
+        """
+        return self.output_dtypes
+
+    def _op_prepare_before_hook_tensor_first_call(
+        self,
+        op: Callable,
+        arg: Any,
+        arg_idx: int,
+        input_observed_arg_idxs: Optional[List[int]],
+        arg_tensor_infos: List[Optional[QTensorInfo]],
+        func_output_dtype_type: FuncOutputDTypeType,
+        qtensor_id: List[int],
+    ) -> None:
+        # TODO(next): fix this for torch.cat
+        if not isinstance(arg, torch.Tensor):
+            arg_tensor_infos.append(None)
+            return
+
+        # If a tensor does not have an ID, add it. This allows
+        # us to track inputs shared by multiple quantizeable modules.
+        if not hasattr(arg, '_qtensor_info'):
+            arg._qtensor_info = QTensorInfo(qtensor_id[0], arg.dtype)  # type: ignore[attr-defined]
+            qtensor_id[0] += 1
+        arg_tensor_infos.append(arg._qtensor_info)  # type: ignore[attr-defined]
+
+        if input_observed_arg_idxs is not None and \
+                arg_idx not in input_observed_arg_idxs:
+            return
+
+        if func_output_dtype_type == FuncOutputDTypeType.DTYPE_DEPENDS_ON_QCONFIG:
+            # if the existing inf_dtype is not torch.quint8, add an observer
+            # which will be converted to a quant later
+            # TODO(future PR): share these observers if multiple ops need
+            # this quant.
+            # TODO(future PR): create from qconfig of op instead of global
+            # qconfig.
+            if arg._qtensor_info.inf_dtype != torch.quint8:  # type: ignore[attr-defined]
+                tensor_id = arg._qtensor_info.id  # type: ignore[attr-defined]
+                weight_arg_idx = get_weight_arg_idx(op)
+                obs = self.qconfig.weight() if arg_idx == weight_arg_idx else \
+                    self.qconfig.activation()
+                self.tensor_id_to_observer[str(tensor_id)] = obs
+
+    def op_prepare_before_hook(
```

Comment:
the name is non-descriptive on purpose (to prevent leaking quantization logic into auto_quant_state.py), but agreed, I can make it easier to read.  Would it work if this function called well named helper functions?

```
def op_prepare_before_hook(...):
    if first_call:
      self._build_subgraphs(...)
      ...
    else:
      self._observe_args_kwargs(...)
      ...
```

Reply:
I refactored the first_call logic into a separate function, hope this helps

Reply:
yeah this works too, thanks!